File: 1402-reducing-dishes/1402-reducing-dishes.cpp
class Solution {
public:
    int helper(vector<int>& arr, int idx, int time, vector<vector<int>>& dp) {
        if(idx == arr.size()) {
            return dp[idx][time] = 0;
        }
        
        if(dp[idx][time] != -1) {
            return dp[idx][time];
        }
        
        int ans = max(helper(arr, idx + 1, time, dp), helper(arr, idx + 1, time + 1, dp) + arr[idx] * time);
        return dp[idx][time] = ans;
    }
    
    int maxSatisfaction(vector<int>& arr) {
        sort(arr.begin(), arr.end());
        vector<vector<int>> dp(arr.size() + 1, vector<int>(arr.size() + 2, -1));
        int ans = helper(arr, 0, 1, dp);
        return ans;
    }
};

File: 1402-reducing-dishes/1402-reducing-dishes.java
class Solution {
    public int memo(int[] satisfaction, int idx, int time, int[][] dp) {
        if (idx ==  satisfaction.length) {
            return 0;
        } 
        
        if (dp[idx][time] != -1) {
            return dp[idx][time];
        }
        
        int inc = memo(satisfaction, idx + 1, time + 1, dp) + satisfaction[idx] * time;
        int exc = memo(satisfaction, idx + 1, time, dp);
        
        return dp[idx][time] = Math.max(inc, exc);
    }
    
    public int maxSatisfaction(int[] satisfaction) {
        int n = satisfaction.length;
        
        Arrays.sort(satisfaction);
        
        int[][] dp = new int[n][n + 1];
        for (int[] a : dp) {
            Arrays.fill(a, -1);
        }
        
        return memo(satisfaction, 0, 1, dp);
    }
}